test/unit_test/runtime/magnitude_test.cpp

```diff
@@ -385,27 +262,79 @@ TEST_CASE("is_prime detects primes")
   }
 }
 
+TEST_CASE("is_valid_base_power")
+{
+  SECTION("0 power is invalid") {
+    REQUIRE(is_valid_base_power(base_power{2}));
+    CHECK(!is_valid_base_power(base_power{2, 0}));
+
+    REQUIRE(is_valid_base_power(base_power{41}));
+    CHECK(!is_valid_base_power(base_power{41, 0}));
+
+    REQUIRE(is_valid_base_power(base_power<pi_base>{}));
+    CHECK(!is_valid_base_power(base_power<pi_base>{0}));
+  }
+
+  SECTION("non-prime integers are invalid") {
+    CHECK(!is_valid_base_power(base_power{-8}));
+    CHECK(!is_valid_base_power(base_power{0}));
+    CHECK(!is_valid_base_power(base_power{1}));
+
+    CHECK(is_valid_base_power(base_power{2}));
+    CHECK(is_valid_base_power(base_power{3}));
+
+    CHECK(!is_valid_base_power(base_power{4}));
+  }
+
+  SECTION("non-positive floating point bases are invalid") {
+    CHECK(!is_valid_base_power(base_power<invalid_zero_base>{}));
+    CHECK(!is_valid_base_power(base_power<invalid_negative_base>{}));
+  }
+}
+
 TEST_CASE("pairwise_all evaluates all pairs")
```

Comment:
@mpusz Have you thought about white-box testing in a modules world? I personally use a `.details` module which gets the same treatment as public API modules in terms of tests.

Reply:
Actually, I do not have much experience with C++20 modules so far. If you have any then I will be glad to learn and benefit from your suggestions. The only action point I thought of so far is to try to move all `namespace detail` implementation to separate header files in the `bits` directory so it will be easier to define exported and private module interfaces.

Reply:
I move things to a common `.details` module only when they're used across different modules. For something like `pairwise_all`, I choose to leave it in the only module it is used and test it indirectly through the public API. In this case, besides testing `magnitude` (the to-be replacement of `ratio`) normally, I'd add additional test cases, if necessary, to prove that its use of `pairwise_all` works as expected.

Reply:
Does it mean that you use modules already? I thought that they are ready only on MSVC and if I remember correctly you do not use this compiler. Also, there is still no support for them in CMake.

If modules are already "usable" on major compilers we should add support for them and check how it addresses compile-time issues for the SI definitions.

Reply:
I use https://github.com/JohelEGP/jegp.cmake_modules/#project-modules to some amount of success. There are many bugs. GCC is particularly plagued with bugs relating to the C++ Standard Library. Can't wait to have `std` as a module.